diff --git a/ThirdPart/src/LIBMEDCOUPLING/MEDCouplingMemArray.cxx b/ThirdPart/src/LIBMEDCOUPLING/MEDCouplingMemArray.cxx
index 9e7408f2..daa545f4 100644
--- a/ThirdPart/src/LIBMEDCOUPLING/MEDCouplingMemArray.cxx
+++ b/ThirdPart/src/LIBMEDCOUPLING/MEDCouplingMemArray.cxx
@@ -111,6 +111,7 @@ void DataArrayDouble::FindTupleIdsNearTuplesAlg(const BBTreePts<SPACEDIM,int>& m
 template<int SPACEDIM>
 void DataArrayDouble::FindClosestTupleIdAlg(const BBTreePts<SPACEDIM,int>& myTree, double dist, const double *pos, int nbOfTuples, const double *thisPt, int thisNbOfTuples, int *res)
 {
+  if (dist==0) dist=1.e-12;
   double distOpt(dist);
   const double *p(pos);
   int *r(res);
