diff --git a/src/ParaMEDMEM/ElementLocator.cxx b/src/ParaMEDMEM/ElementLocator.cxx
index 31d7033cd..ee9d4d0c8 100644
--- a/src/ParaMEDMEM/ElementLocator.cxx
+++ b/src/ParaMEDMEM/ElementLocator.cxx
@@ -267,30 +267,31 @@ namespace MEDCoupling
     distant_mesh_tmp->resizeForUnserialization(tinyInfoDistant,v1Distant,v2Distant,unusedTinyDistantSts);
     mcIdType nbLocalElems=0;
     mcIdType nbDistElem=0;
-    mcIdType *ptLocal=0;
+    const mcIdType *ptLocal=0;
     mcIdType *ptDist=0;
     if(v1Local)
       {
         nbLocalElems=v1Local->getNbOfElems();
-        ptLocal=v1Local->getPointer();
+        ptLocal=v1Local->getConstPointer();
       }
     if(v1Distant)
       {
         nbDistElem=v1Distant->getNbOfElems();
         ptDist=v1Distant->getPointer();
       }
-    comm_interface.sendRecv(ptLocal, (int)nbLocalElems, MPI_ID_TYPE,
+    mcIdType * ptLocalNonConst = const_cast<mcIdType *>(ptLocal);
+    comm_interface.sendRecv(ptLocalNonConst, (int)nbLocalElems, MPI_ID_TYPE,
                             iprocdistant_in_union, 1111,
                             ptDist, (int)nbDistElem, MPI_ID_TYPE,
                             iprocdistant_in_union,1111,
                             *comm, &status);
     nbLocalElems=0;
-    double *ptLocal2=0;
+    const double *ptLocal2=0;
     double *ptDist2=0;
     if(v2Local)
       {
         nbLocalElems=v2Local->getNbOfElems();
-        ptLocal2=v2Local->getPointer();
+        ptLocal2=v2Local->getConstPointer();
       }
     nbDistElem=0;
     if(v2Distant)
@@ -298,7 +299,8 @@ namespace MEDCoupling
         nbDistElem=v2Distant->getNbOfElems();
         ptDist2=v2Distant->getPointer();
       }
-    comm_interface.sendRecv(ptLocal2, (int)nbLocalElems, MPI_DOUBLE,
+    double * ptLocal2NonConst = const_cast<double *>(ptLocal2);
+    comm_interface.sendRecv(ptLocal2NonConst, (int)nbLocalElems, MPI_DOUBLE,
                             iprocdistant_in_union, 1112,
                             ptDist2, (int)nbDistElem, MPI_DOUBLE,
                             iprocdistant_in_union, 1112, 
