diff --git a/src/MEDCoupling/CMakeLists.txt b/src/MEDCoupling/CMakeLists.txt
index 04e0b5b3..b8d9e5d1 100644
--- a/src/MEDCoupling/CMakeLists.txt
+++ b/src/MEDCoupling/CMakeLists.txt
@@ -33,6 +33,7 @@ INCLUDE_DIRECTORIES(
   ${CMAKE_CURRENT_SOURCE_DIR}/../INTERP_KERNEL/Geometric2D
   ${CMAKE_CURRENT_SOURCE_DIR}/../INTERP_KERNEL/ExprEval
   ${CMAKE_CURRENT_SOURCE_DIR}/../INTERP_KERNEL/GaussPoints
+  ${CMAKE_CURRENT_SOURCE_DIR}/../ICoCo
   )
 
 SET(medcoupling_SOURCES
@@ -72,7 +73,7 @@ SET(medcoupling_SOURCES
   MEDCouplingPartDefinition.cxx
   MEDCouplingSkyLineArray.cxx
   MEDCouplingVoronoi.cxx
-  ../ICoCo/ICoCoField.cxx
+  ../ICoCo/ICoCoField.cpp      # [ABN] Yes, .cpp, this is imposed by ICoCo.
   ../ICoCo/ICoCoMEDField.cxx
   )
 
@@ -89,7 +90,9 @@ ADD_LIBRARY(medcouplingremapper ${medcouplingremapper_SOURCES})
 TARGET_LINK_LIBRARIES(medcouplingremapper medcouplingcpp)
 INSTALL(TARGETS medcouplingremapper EXPORT ${PROJECT_NAME}TargetGroup DESTINATION ${MEDCOUPLING_INSTALL_LIBS})
 
-FILE(GLOB medcoupling_HEADERS_HXX "${CMAKE_CURRENT_SOURCE_DIR}/*.hxx")
+FILE(GLOB medcoupling_HEADERS_HXX "${CMAKE_CURRENT_SOURCE_DIR}/*.hxx" 
+                                  "${CMAKE_CURRENT_SOURCE_DIR}/../ICoCo/*.hxx"
+                                  "${CMAKE_CURRENT_SOURCE_DIR}/../ICoCo/*.h")
 FILE(GLOB medcoupling_HEADERS_TXX "${CMAKE_CURRENT_SOURCE_DIR}/*.txx")
 FILE(GLOB icoco_HEADERS_HXX "${CMAKE_CURRENT_SOURCE_DIR}/../ICoCo/*.hxx")
 INSTALL(FILES ${medcoupling_HEADERS_HXX} ${medcoupling_HEADERS_TXX} ${icoco_HEADERS_HXX} MEDCouplingNatureOfFieldEnum DESTINATION ${MEDCOUPLING_INSTALL_HEADERS})
diff --git a/src/MEDCoupling_Swig/CMakeLists.txt b/src/MEDCoupling_Swig/CMakeLists.txt
index 8dd9feb7..89f92135 100644
--- a/src/MEDCoupling_Swig/CMakeLists.txt
+++ b/src/MEDCoupling_Swig/CMakeLists.txt
@@ -36,7 +36,9 @@ SET (MEDCoupling_SWIG_DPYS_FILES
     MEDCouplingMemArray.i
     MEDCouplingFieldDiscretization.i
     MEDCouplingFinalize.i
-    MEDCouplingTypemaps.i)
+    MEDCouplingTypemaps.i
+    ../ICoCo/Swig/ICoCoMEDField.i
+)
 
 SET (MC_pyTestFiles
     ${ALL_TESTS}
@@ -56,6 +58,7 @@ SET (MC_Swig_interf
     MEDCouplingTypemaps.i
     MEDCouplingDataArrayTypemaps.i
     MEDCouplingDataArrayTraits.hxx
+    ../ICoCo/Swig/ICoCoMEDField.i
 )
 
 INCLUDE_DIRECTORIES(
@@ -70,6 +73,8 @@ INCLUDE_DIRECTORIES(
   ${CMAKE_CURRENT_SOURCE_DIR}/../INTERP_KERNEL/Geometric2D
   ${CMAKE_CURRENT_SOURCE_DIR}/../INTERP_KERNEL/ExprEval
   ${CMAKE_CURRENT_SOURCE_DIR}/../INTERP_KERNEL/GaussPoints
+  ${CMAKE_CURRENT_SOURCE_DIR}/../ICoCo
+  ${CMAKE_CURRENT_SOURCE_DIR}/../ICoCo/Swig
   ${PROJECT_BINARY_DIR}/doc
   )
 
diff --git a/src/MEDCoupling_Swig/MEDCoupling.i b/src/MEDCoupling_Swig/MEDCoupling.i
index 57ae1cd4..9f44e9e4 100644
--- a/src/MEDCoupling_Swig/MEDCoupling.i
+++ b/src/MEDCoupling_Swig/MEDCoupling.i
@@ -24,6 +24,7 @@
 #endif
 
 %include "MEDCouplingCommon.i"
+%include "ICoCoMEDField.i"
 
 %pythoncode %{
 def MEDCouplingDataArrayDoubleIadd(self,*args):
diff --git a/src/PyWrapping/CMakeLists.txt b/src/PyWrapping/CMakeLists.txt
index 56c928ff..2a93dac7 100644
--- a/src/PyWrapping/CMakeLists.txt
+++ b/src/PyWrapping/CMakeLists.txt
@@ -53,6 +53,7 @@ INCLUDE_DIRECTORIES(
   ${CMAKE_CURRENT_SOURCE_DIR}/../MEDPartitioner
   ${CMAKE_CURRENT_SOURCE_DIR}/../MEDPartitioner_Swig
   ${CMAKE_CURRENT_SOURCE_DIR}/../ICoCo
+  ${CMAKE_CURRENT_SOURCE_DIR}/../ICoCo/Swig
   ${CMAKE_CURRENT_SOURCE_DIR}/../ParaMEDMEM
   ${CMAKE_CURRENT_SOURCE_DIR}/../ParaMEDMEM_Swig
   ${PROJECT_BINARY_DIR}/doc
diff --git a/src/PyWrapping/medcoupling.i b/src/PyWrapping/medcoupling.i
index 72032a3a..4bf4ab1b 100644
--- a/src/PyWrapping/medcoupling.i
+++ b/src/PyWrapping/medcoupling.i
@@ -30,6 +30,8 @@
 
 %include "MEDCouplingRemapperCommon.i"
 
+%include "ICoCoMEDField.i"
+
 #ifdef WITH_MED_FILE
 %include "MEDLoaderCommon.i"
 #endif
