diff -u Interpolation2D3D.txx Interpolation2D3D.txx
--- Interpolation2D3D.txx	2022-11-22 12:19:01.000000000 +0100
+++ Interpolation2D3D.txx	2023-03-21 12:54:45.822888302 +0100
@@ -99,7 +99,8 @@
 
     // create BBTree structure
     // [ABN] Adjust 2D bounding box (those might be flat in the cases where the 2D surf are perfectly aligned with the axis)
-    BBTreeStandAlone<3,ConnType> tree( BuildBBTreeWithAdjustment(srcMesh,[this,&intersector](double *bbox, typename MyMeshType::MyConnType sz){ this->performAdjustmentOfBB(intersector.get(),bbox,sz); }) );
+    std::function<void(double *,typename MyMeshType::MyConnType)> lambda = [this,&intersector](double *bbox, typename MyMeshType::MyConnType sz){ this->performAdjustmentOfBB(intersector.get(),bbox,sz); };
+    BBTreeStandAlone<3,ConnType> tree( BuildBBTreeWithAdjustment(srcMesh, lambda) );
 
     // for each target element, get source elements with which to calculate intersection
     // - calculate intersection by calling intersectCells
diff -u Interpolation3D1D.txx Interpolation3D1D.txx
--- Interpolation3D1D.txx	2022-11-22 12:19:01.000000000 +0100
+++ Interpolation3D1D.txx	2023-03-21 12:54:45.822888302 +0100
@@ -68,7 +68,8 @@
     // create empty maps for all source elements
     result.resize(intersector->getNumberOfRowsOfResMatrix());
 
-    BBTreeStandAlone<3,ConnType> tree( BuildBBTreeWithAdjustment(srcMesh,[this](double *bbox, typename MyMeshType::MyConnType sz){ this->adjustBoundingBoxes(bbox,sz); }) );
+    std::function<void(double *,typename MyMeshType::MyConnType)> lambda = [this](double *bbox, typename MyMeshType::MyConnType sz){ this->adjustBoundingBoxes(bbox,sz); };
+    BBTreeStandAlone<3,ConnType> tree( BuildBBTreeWithAdjustment(srcMesh,lambda) );
 
     // for each target element, get source elements with which to calculate intersection
     // - calculate intersection by calling intersectCells
